crates/biome_formatter/src/trivia.rs

```diff
@@ -1,15 +1,44 @@
 //! Provides builders for comments and skipped token trivia.
 
+use crate::comments::is_doc_comment;
 use crate::format_element::tag::VerbatimKind;
 use crate::prelude::*;
 use crate::{
     comments::{CommentKind, CommentStyle},
     write, Argument, Arguments, CstFormatContext, FormatRefWithRule, GroupId, SourceComment,
     TextRange,
 };
-use biome_rowan::{Language, SyntaxNode, SyntaxToken};
+use biome_rowan::{Language, SyntaxNode, SyntaxToken, TextSize};
 #[cfg(debug_assertions)]
 use std::cell::Cell;
+use std::ops::Sub;
+
+/// Returns true if:
+/// - `next_comment` is Some, and
+/// - both comments are documentation comments, and
+/// - both comments are multiline, and
+/// - the two comments are immediately adjacent to each other, with no characters between them.
+///
+/// In this case, the comments are considered "nestled" - a pattern that JSDoc uses to represent
+/// overloaded types, which get merged together to create the final type for the subject. The
+/// comments must be kept immediately adjacent after formatting to preserve this behavior.
+///
+/// There isn't much documentation about this behavior, but it is mentioned on the JSDoc repo
+/// for documentation: https://github.com/jsdoc/jsdoc.github.io/issues/40. Prettier also
+/// implements the same behavior: https://github.com/prettier/prettier/pull/13445/files#diff-3d5eaa2a1593372823589e6e55e7ca905f7c64203ecada0aa4b3b0cdddd5c3ddR160-R178
+fn should_nestle_adjacent_doc_comments<L: Language>(
+    first_comment: &SourceComment<L>,
+    second_comment: &SourceComment<L>,
+) -> bool {
+    let first = first_comment.piece();
+    let second = second_comment.piece();
+
+    first.has_newline()
+        && second.has_newline()
+        && (second.text_range().start()).sub(first.text_range().end()) == TextSize::from(0)
+        && is_doc_comment(first)
+        && is_doc_comment(second)
```

Comment:
I tried ordering this to make it do the least work possible. I imagine the text range stuff gets inlined into just integer comparisons and a subtraction by the compiler, but i'm not totally sure. Probably doesn't really matter since this function is unlikely to be called in most cases anyway.